plonky2/src/util/serialization.rs

```diff
@@ -20,63 +23,83 @@ use crate::plonk::proof::{
     CompressedProof, CompressedProofWithPublicInputs, OpeningSet, Proof, ProofWithPublicInputs,
 };
 
-#[derive(Debug)]
-pub struct Buffer(Cursor<Vec<u8>>);
```

Comment:
Instead of using `Vec<u8>` as a write-only alternative alongside `Buffer`, why don't we change `Buffer` to not rely on `Cursor`? Seems cleaner IMO to have have a single buffer object, and it'd enable deserialization without std. We could also remove the `Result`s if we're no longer using `Cursor`.

Reply:
For this PR I wanted to do the minimal changes to the serialization API that resulted in `no-std` compatibility, in this case, we can get `serialization` for free (but not `deserialization`). We can of course expand on this minimal version to support everything, but then I would recommend using `serde` or `bincode` instead of building your own serialization library. Also, another point to be made is that the implementation of (de)serialization of a particular object should be local to that object instead of local to the (de)serializer. I didn't want to fix this to reduce the scope of this PR, but I think that would also be a great change to make. We can maybe move this to another serialization-themed Issue/PR.

> Instead of using Vec<u8> as a write-only alternative alongside Buffer, why don't we change Buffer to not rely on Cursor? Since Cursor just adds a position it seems easy to DIY.

Are you saying to implement a `no-std` compliant `Cursor` ourselves in `util`?

> Seems cleaner IMO to have have a single buffer object, and it'd enable deserialization without std. We could also remove the Results if we're no longer using Cursor.

I don't think you can ever remove `Result` for deserialization (serialization you can in some circumstances like if you're only doing in-memory serialization). For deserialization you are going from unstructured to structured data, so its not just that the conversion may fail from some I/O or resource error but that the data is actually not structured correctly, like there not being enough bytes for a field element; that's not an I/O error that's an error in the encoding directly.

Reply:
Good point, `Result` makes sense at least for deserialization.

We used serde with CBOR in the past, but there were some sub-optimal things we wanted to optimize. I think the main thing was that we wanted to remove length prefixes from `Vec`s whose lengths can be inferred from `CircuitConfig`. With serde I don't think we could access `CircuitConfig` from deserialization code (but correct me if wrong).

My main concern with the code here is that (if std is enabled) we now have two ways to serialize a proof, one with `Buffer` and one with `Vec<u8>`. I guess the easiest solution is to just not `impl Write for Buffer`.

Reply:
It's just a minor thing so we could also leave it for now, and later change `Buffer` to work without Cursor, and at that point remove the `Vec<u8>` writer. WDYT?

Reply:
> We used serde with CBOR in the past, but there were some sub-optimal things we wanted to optimize. I think the main thing was that we wanted to remove length prefixes from `Vec`s whose lengths can be inferred from `CircuitConfig`. With serde I don't think we could access `CircuitConfig` from deserialization code (but correct me if wrong).

Trying to remove length prefixes is a property of the "(de)serializer" (i.e. encoding format) rather than the "(de)serialized object" so if you wanted this you would create a custom `serde::Serializer` and `serde::Deserializer` implementation that can take a `CircuitConfig` as input (see [`serde_json::ser::Serializer`](https://docs.rs/serde_json/latest/serde_json/ser/struct.Serializer.html) as an example: we can specify the pretty-print by adding a custom [`Formatter`](https://docs.rs/serde_json/latest/serde_json/ser/trait.Formatter.html) implementation). This is very do-able.

> My main concern with the code here is that (if `std` is enabled) we now have two ways to serialize a proof, one with `Buffer` and one with `Vec<u8>`. I guess the easiest solution is to just not `impl Write for Buffer`.

I think this is fine for a few reasons:
1. Abstract interface for writing (if you don't know which writer you want just use the trait)
2. User flexibility if they want serialization and deserialization or just serialization.

I agree it is suboptimal that we don't have deserialization in `no-std` but I don't want to add more complexity to this PR.

> It's just a minor thing so we could also leave it for now, and later change `Buffer` to work without std Cursor, and at that point remove the `Vec<u8>` writer. WDYT?

Yeah, I think that's the best option for engineering complexity right now, we can implement a minimal cursor API (or see if we can get rid of position dependence entirely).

Reply:
Sounds good, let's leave it for now and maybe add that Cursor-like functionality later.

Hm I hadn't thought about a custom `Serializer`; I haven't seen a `Serializer` focused on a particular application before. Are you thinking this would handle arbitrary objects, but apply some special optimizations when it encounters a proof object?

Implementing `Serialize`/`Deserialize` seems ideal in terms of user-friendliness, so that proof objects could be embedded in larger objects, which could be serialized with whatever `Serializer` is suitable for the broader application. But it seems like that would prevent accessing `CircuitConfig`.